src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -203,42 +211,148 @@ MakeIntermediateResultHTAB()
 
 /*
  * FindAllWorkerNodesUsingSubplan creates a list of worker nodes that
- * may need to access subplan results.
+ * may need to access subplan results. The function also sets writeToLocalFile
+ * flag if the result should also need be written locally.
  */
 List *
 FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
 							   char *resultId)
 {
-	List *workerNodeList = NIL;
 	IntermediateResultsHashEntry *entry =
 		SearchIntermediateResult(intermediateResultsHash, resultId);
 
+	List *remoteWorkerNodes = FindAllRemoteWorkerNodesUsingSubplan(entry);
+	if (remoteWorkerNodes == NIL)
+	{
+		/*
+		 * This could happen in two cases:
+		 * (a) Subquery in the having
+		 * (b) The intermediate result is not used, such as RETURNING of a
+		 *     modifying CTE is not used
+		 *
+		 * For SELECT, Postgres/Citus is clever enough to not execute the CTE
+		 * if it is not used at all, but for modifications we have to execute
+		 * the queries.
+		 */
+		entry->writeLocalFile = true;
+	}
+
+	/*
+	 * Don't include the current worker if the result will be written to local
+	 * file as this would be very inefficient and potentially leading race
+	 * conditions while tring to write the same file twice.
+	 */
+	if (entry->writeLocalFile)
+	{
+		remoteWorkerNodes = RemoveLocalNodeFromWorkerList(remoteWorkerNodes);
+	}
+
+	LogIntermediateResultMulticastSummary(entry, remoteWorkerNodes);
+
+	return remoteWorkerNodes;
+}
+
+
+/*
+ * FindAllRemoteWorkerNodesUsingSubplan goes over the nodeIdList of the
+ * intermediate result entry, and returns a list of workerNodes that the
+ * entry should be multi-casted to. The aim of the function is to filter
+ * out nodes with LOCAL_NODE_ID.
+ */
+static List *
+FindAllRemoteWorkerNodesUsingSubplan(IntermediateResultsHashEntry *entry)
+{
+	List *workerNodeList = NIL;
+
 	ListCell *nodeIdCell = NULL;
 	foreach(nodeIdCell, entry->nodeIdList)
 	{
-		WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+		uint32 nodeId = lfirst_int(nodeIdCell);
```

Comment:
in `AppendAllAccessedWorkerNodes` we do append `placement->nodeId` which is a `int` and here we read the value as a uint32. I don't think that is what should happen here.

I haven't found any other places where we add nodeId's to `entry->nodeIdList` yet, so I don't know if the bug is in `AppendAllAccessedWorkerNodes` or here. Guess you are better situated to find the correct type used for nodeId's in this list. Can you respond with your thoughts and reasoning after deciding on one or the other.

Reply:
Can we change the sequence that is used to assign `nodeId` so that `UINT32_MAX` will never be assigned and causes an error? I guess we can deduct the end of the range by one during the citus upgrade script.